system/Entity.php

```diff
@@ -359,31 +359,49 @@ public function __set(string $key, $value = null)
 
 		if (! $isNullable || ! is_null($value))
 		{
-			// CSV casts need to be imploded.
-			if ($castTo === 'csv')
+			switch ($castTo)
 			{
-				$value = implode(',', $value);
-			}
-
-			// Array casting requires that we serialize the value
-			// when setting it so that it can easily be stored
-			// back to the database.
-			if ($castTo === 'array')
-			{
-				$value = serialize($value);
-			}
-
-			// JSON casting requires that we JSONize the value
-			// when setting it so that it can easily be stored
-			// back to the database.
-			if (($castTo === 'json' || $castTo === 'json-array') && function_exists('json_encode'))
-			{
-				$value = json_encode($value, JSON_UNESCAPED_UNICODE);
-
-				if (json_last_error() !== JSON_ERROR_NONE)
-				{
-					throw CastException::forInvalidJsonFormatException(json_last_error());
-				}
+				case 'bool':
+				case 'boolean':
+					$value = filter_var($value, FILTER_VALIDATE_BOOLEAN, $isNullable ? FILTER_NULL_ON_FAILURE : 0);
```

Comment:
I don't understand this one. If `__set()` casting is supposed to "mirror" `castAs()` why are we converting back to boolean? Is this to prevent `null` when a cast is not nullable?

It also seems like we are introducing some additional values, "on off yes no":

>Returns true for "1", "true", "on" and "yes". Returns false otherwise.
> If FILTER_NULL_ON_FAILURE is set, false is returned only for "0", "false", "off", "no", and "", and null is returned for all non-boolean values.

Reply:
The problem with not having setter is that it will allow setting any sort of string as value for bool property. Trying to write such strings in many NoSQL databases will result in error

The problem with current casts functionality it that there is no way to limit what is allowed to be written in attributes array.

And the bool is just example how can it produce sorts of issues. Currently I have a custom setter for each bool value in my entities where I'm doing those conversions. 

And the problem gets even more annoying when there's a datamap in play as those setters will have completely different name

Reply:
I like what's suggested in #4308 

It will make this PR completely redundant. And it will give a lot of possibilities with casting so I'd rather #4308 to be merged than this one.